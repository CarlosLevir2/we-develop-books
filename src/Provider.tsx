import { QueryClientProvider } from 'react-query';
import { ThemeProvider } from 'styled-components';
import useTheme from './hooks/theme';
import queryClient from './services/queryClient';

type Props = {
  children: React.ReactNode;
};

export default function Provider({ children }: Props) {
  const { theme } = useTheme();

  return (
    <ThemeProvider theme={theme}>
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    </ThemeProvider>
  );
}
