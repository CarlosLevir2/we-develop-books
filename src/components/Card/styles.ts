import styled from 'styled-components';
import { HiStar } from 'react-icons/hi';

export const Card = styled.li`
  display: flex;
  flex-direction: column;
  background: none;
  padding: 0;
  color: ${({ theme }) => theme.colors.text};

  & {
    margin-left: 20px;
  }
`;

export const CardTitle = styled.span`
  font-size: 14px;
  display: block;
  width: 162px;
  color: ${({ theme }) => theme.colors.text.primary};
  margin-top: 8px;
  text-overflow: ellipsis;
  white-space: nowrap;
  pointer-events: auto;
  overflow: hidden;
  font-weight: bold;
`;

export const CardImage = styled.img`
  font-size: 14px;
  display: block;
  width: 192px;
  color: ${({ theme }) => theme.colors.text};
  margin-top: 8px;
  text-overflow: ellipsis;
  white-space: nowrap;
  pointer-events: auto;
  overflow: hidden;

  border-radius: 8px;
  object-fit: cover;
  object-position: center center;
  box-shadow: 0 1px 1px rgba(53, 71, 121, 0.11),
    0 2px 2px rgba(53, 71, 121, 0.11), 0 4px 4px rgba(53, 71, 121, 0.11),
    0 6px 8px rgba(53, 71, 121, 0.11), 0 8px 16px rgba(53, 71, 121, 0.11);
  width: 160px;
  height: 226px;
`;

export const Information = styled.div`
  display: flex;
  flex-direction: column;
`;

export const Star = styled(HiStar)`
  cursor: pointer;
  > svg + svg {
    margin-left: 4px;
  }
`;

export const ShowDescriptionButton = styled.button`
  border: 0;
  width: 12rem;
  height: 3rem;
  background-color: ${({ theme }) => theme.colors.accent.primary};
  border-radius: 0.4rem;
  color: ${({ theme }) => theme.colors.text.highlight};
  cursor: pointer;
`;

export const Description = styled.span`
  font-size: 14px;
  display: block;
  width: 162px;
  color: ${({ theme }) => theme.colors.text.primary};
  margin-top: 8px;
  text-overflow: ellipsis;
  white-space: nowrap;
  pointer-events: auto;
  overflow: hidden;
  font-weight: bold;
`;
