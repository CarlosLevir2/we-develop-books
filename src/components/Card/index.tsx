import { useCallback, useMemo, useState } from 'react';
import useTheme from '../../hooks/theme';
import { Manga } from '../../services/requests/types';
import { useFavoriteMangas } from '../../store/useFavoriteMangas';

import * as S from './styles';

type Props = {
  manga: Manga;
  showFavorites?: boolean;
};

const Card = ({ manga, showFavorites = false }: Props) => {
  const [showDescription, setShowDescription] = useState(false);
  const { favorites } = useFavoriteMangas();
  const { theme } = useTheme();

  const isFavorite = useMemo(
    () => favorites?.includes(manga.mal_id),
    [favorites, manga.mal_id]
  );

  const toggleFavoriteManga = useCallback(() => {
    const shouldRemove = favorites?.includes(manga.mal_id);

    const newFavorites = shouldRemove
      ? favorites?.filter((favorite) => favorite !== manga.mal_id)
      : [...(favorites || []), manga.mal_id];

    useFavoriteMangas.setState({ favorites: newFavorites });
  }, [favorites, manga.mal_id]);

  const toggleDescription = useCallback(() => {
    setShowDescription((oldShowDescription) => !oldShowDescription);
  }, []);

  if (showFavorites && !isFavorite) {
    return null;
  }

  return (
    <S.Card>
      <S.CardImage src={manga.image_url} alt={manga.title} />
      <S.Information>
        <S.Star
          size={28}
          color={theme.colors.stars[isFavorite ? 'fill' : 'empty']}
          onClick={toggleFavoriteManga}
        />
        <S.CardTitle>{manga.title}</S.CardTitle>

        <div>
          <S.ShowDescriptionButton onClick={toggleDescription}>
            Show Description
          </S.ShowDescriptionButton>
          {showDescription && <S.Description>{manga.url}</S.Description>}
        </div>
      </S.Information>
    </S.Card>
  );
};
export default Card;
