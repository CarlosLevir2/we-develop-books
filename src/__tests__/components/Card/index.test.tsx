import { render, screen } from '@testing-library/react';
import Card from '../../../components/Card';
import Provider from '../../../Provider';
import { mangasMock } from '../../__mocks__/mangas';

const firstManga = mangasMock.top[0];

test('renders card component', () => {
  render(
    <Provider>
      <Card manga={firstManga} />
    </Provider>
  );
  const linkElement = screen.getByText(firstManga.title);
  expect(linkElement).toBeInTheDocument();
});

test('match snapshot', () => {
  const { container } = render(
    <Provider>
      <Card manga={firstManga} />
    </Provider>
  );
  expect(container).toMatchSnapshot();
});
