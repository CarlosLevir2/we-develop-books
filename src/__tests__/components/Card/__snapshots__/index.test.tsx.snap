// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match snapshot 1`] = `
<div>
  <li
    class="sc-bdvvtL eGdjyq"
  >
    <img
      alt="Berserk"
      class="sc-dkPtRN dfuFxG"
      src="https://cdn.myanimelist.net/images/manga/1/157897.jpg?s=f03b5f8bfeb0b0962b7d5e7cb9a8d0d3"
    />
    <div
      class="sc-hKwDye bSWfzB"
    >
      <svg
        class="sc-eCImPb imapUP"
        color="#6e727a"
        fill="currentColor"
        height="28"
        stroke="currentColor"
        stroke-width="0"
        style="color: rgb(110, 114, 122);"
        viewBox="0 0 20 20"
        width="28"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.049 2.927c.3-.921 1.603-.921 1.902 0l1.07 3.292a1 1 0 00.95.69h3.462c.969 0 1.371 1.24.588 1.81l-2.8 2.034a1 1 0 00-.364 1.118l1.07 3.292c.3.921-.755 1.688-1.54 1.118l-2.8-2.034a1 1 0 00-1.175 0l-2.8 2.034c-.784.57-1.838-.197-1.539-1.118l1.07-3.292a1 1 0 00-.364-1.118L2.98 8.72c-.783-.57-.38-1.81.588-1.81h3.461a1 1 0 00.951-.69l1.07-3.292z"
        />
      </svg>
      <span
        class="sc-gsDKAQ bcJVxq"
      >
        Berserk
      </span>
      <div>
        <button
          class="sc-jRQBWg ioDimu"
        >
          Show Description
        </button>
      </div>
    </div>
  </li>
</div>
`;
